pkg/sst/sst.go

```diff
@@ -234,3 +241,243 @@ func getBits(val, i, j uint32) uint32 {
 func isBitSet(val, n uint32) bool {
 	return val&(1<<n) != 0
 }
+
+func setBit(val, n uint32) uint32 {
+	return val | (1 << n)
+}
+
+func clearBit(val, n uint32) uint32 {
+	return val &^ (1 << n)
+}
+
+func checkBFSupportStatus(info *SstPackageInfo) error {
+	if !info.PPSupported {
+		sstlog.Debugf("SST PP feature not supported, only profile level %d is valid", info.PPCurrentLevel)
+	}
+
+	if !info.BFSupported {
+		sstlog.Debugf("SST BF not supported")
+		return fmt.Errorf("SST BF not supported")
+	}
+
+	return nil
+}
+
+func setBFStatus(info *SstPackageInfo, status bool) error {
+	rsp, err := sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_GET_TDP_CONTROL, uint32(info.PPCurrentLevel))
+	if err != nil {
+		return fmt.Errorf("failed to read SST status: %v", err)
+	}
+
+	var req uint32
+
+	if status {
+		req = setBit(rsp, 17)
+	} else {
+		req = clearBit(rsp, 17)
+	}
+
+	if _, err = sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_SET_TDP_CONTROL, req); err != nil {
+		return fmt.Errorf("failed to enable SST %s: %v", "BF", err)
+	}
+
+	info.BFEnabled = status
+
+	return nil
+}
+
+func setMinFreq2Max(cpu int) error {
+	str := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", cpu)
+
+	raw, err := ioutil.ReadFile(str)
+	if err != nil {
+		return err
+	}
+
+	var maxFreq int
+
+	maxFreq, err = strconv.Atoi(strings.TrimSpace(string(raw)))
+	if err != nil {
+		return err
+	}
+
+	minFreqStr := fmt.Sprintf("%d", maxFreq)
+
+	str = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_min_freq", cpu)
+
+	err = ioutil.WriteFile(str, []byte(minFreqStr), 0644)
+	if err != nil {
+		return err
+	}
+
+	//sstlog.Debugf("sst-bf cpu %d min freq %s", cpu, minFreqStr)
+
+	return nil
+}
+
+func getScalingMinFreq(cpu int) (int, error) {
+	str := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_min_freq", cpu)
+
+	raw, err := ioutil.ReadFile(str)
+	if err != nil {
+		return 0, err
+	}
+
+	minScalingFreq, err := strconv.Atoi(strings.TrimSpace(string(raw)))
+	if err != nil {
+		return 0, err
+	}
+
+	//sstlog.Debugf("sst-bf cpu %d scaling min freq %d", cpu, minScalingFreq)
+
+	return minScalingFreq, nil
+}
+
+func getBaseFreq(cpu int) (int, error) {
+	str := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/base_frequency", cpu)
+
+	raw, err := ioutil.ReadFile(str)
+	if err != nil {
+		return 0, err
+	}
+
+	baseFreq, err := strconv.Atoi(strings.TrimSpace(string(raw)))
+	if err != nil {
+		return 0, err
+	}
+
+	//sstlog.Debugf("sst-bf cpu %d base freq %d", cpu, baseFreq)
+
+	return baseFreq, nil
+}
+
+func setScalingMinFreq(cpu int, freq int) error {
+	str := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_min_freq", cpu)
+
+	freqStr := fmt.Sprintf("%d", freq)
+
+	err := ioutil.WriteFile(str, []byte(freqStr), 0644)
+	if err != nil {
+		return err
+	}
+
+	//sstlog.Debugf("sst-bf cpu %d scaling min freq %d", cpu, freqStr)
+
+	return nil
+}
```

Comment:
Looks like a lot of code duplication in these get/set cpufreq attributes. Maybe be could have two helper functions, one for getting and one for setting?

Reply:
Yep, makes sense, will change